src/overlays/effects/ovl_Effect_Ss_Bubble/z_eff_ss_bubble.c

```diff
@@ -5,23 +5,103 @@
+
+    {
+        TexturePtr tex = (Rand_ZeroOne() < 0.5f) ? gEffBubble1Tex : gEffBubble2Tex;
+
+        this->gfx = VIRTUAL_TO_PHYSICAL(SEGMENTED_TO_VIRTUAL(tex));
+    }
+
+    this->pos.x = ((Rand_ZeroOne() - 0.5f) * initParams->xzPosRandScale) + initParams->pos.x;
+    this->pos.y = (((Rand_ZeroOne() - 0.5f) * initParams->yPosRandScale) + initParams->yPosOffset) + initParams->pos.y;
+    this->pos.z = ((Rand_ZeroOne() - 0.5f) * initParams->xzPosRandScale) + initParams->pos.z;
+    Math_Vec3f_Copy(&this->vec, &this->pos);
+    this->life = 1;
+    this->rScale = (((Rand_ZeroOne() * 0.5f) + 1.0f) * initParams->scale) * 100.0f;
+    this->draw = EffectSsBubble_Draw;
+    this->update = EffectSsBubble_Update;
+
+    return 1;
+}
+
+void EffectSsBubble_Draw(GlobalContext* globalCtx, u32 index, EffectSs* this) {
+    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
+    f32 scale = this->rScale / 100.0f;
+
+    OPEN_DISPS(gfxCtx);
+
+    Matrix_InsertTranslation(this->pos.x, this->pos.y, this->pos.z, MTXMODE_NEW);
+    Matrix_Scale(scale, scale, scale, MTXMODE_APPLY);
+
+    gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    func_8012C28C(gfxCtx);
+    gDPSetPrimColor(POLY_OPA_DISP++, 0, 0, 255, 255, 255, 255);
+    gDPSetEnvColor(POLY_OPA_DISP++, 150, 150, 150, 0);
+    gSPSegment(POLY_OPA_DISP++, 0x08, this->gfx);
+    gSPDisplayList(POLY_OPA_DISP++, gEffBubbleDL);
+
+    CLOSE_DISPS(gfxCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Effect_Ss_Bubble/EffectSsBubble_Init.s")
+void EffectSsBubble_Update(GlobalContext* globalCtx2, u32 index, EffectSs* this) {
+    WaterBox* waterBox;
+    f32 waterSurfaceY = this->pos.y;
+    Vec3f ripplePos;
+    GlobalContext* globalCtx = globalCtx2;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Effect_Ss_Bubble/EffectSsBubble_Draw.s")
+    if (!WaterBox_GetSurface1(globalCtx, &globalCtx->colCtx, this->pos.x, this->pos.z, &waterSurfaceY, &waterBox)) {
+        this->life = -1;
+        return;
+    }
+    if (waterSurfaceY < this->pos.y) {
+        ripplePos.x = this->pos.x;
+        ripplePos.y = waterSurfaceY;
+        ripplePos.z = this->pos.z;
+        EffectSsGRipple_Spawn(globalCtx, &ripplePos, 0, 80, 0);
+        this->life = -1;
+    label:
+        return;
+    }
+    if (((globalCtx->gameplayFrames + index) % 8) == 0) {
+        CollisionPoly* colPoly;
+        u32 speed;
+        s16 direction;
+        f32 rVecAdjAmplitude;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Effect_Ss_Bubble/EffectSsBubble_Update.s")
+        BgCheck_EntityRaycastFloor2_1(globalCtx, &globalCtx->colCtx, &colPoly, &this->pos);
+        speed = SurfaceType_GetConveyorSpeed(&globalCtx->colCtx, colPoly, 0x32);
+        if ((speed != 0) && !SurfaceType_GetConveyorType(&globalCtx->colCtx, colPoly, 0x32)) {
+            direction = SurfaceType_GetConveyorDirection(&globalCtx->colCtx, colPoly, 0x32) << 0xA;
```

Comment:
```suggestion
        if ((speed != 0) && !SurfaceType_GetConveyorType(&globalCtx->colCtx, colPoly, BGCHECK_SCENE)) {
            direction = SurfaceType_GetConveyorDirection(&globalCtx->colCtx, colPoly, BGCHECK_SCENE) << 0xA;
```
These are checking the surfaces of the scene